### PR TITLE
Add SELinux support

### LINKED ISSUE #124: SELinux support
We should detect SELinux in our `linux-multi` planner and add steps similar to those found in https://github.com/nix-community/nix-installers/tree/master/selinux.

### PR DESCRIPTION
**Pull request by Hoverbear**
##### Description

Add SELinux support!

Fixes #124 

SELinux users can test this with:

```
curl --proto '=https' --tlsv1.2 -sSf -L https://install.determinate.systems/nix/pr/465 | sh -s -- install
```

##### Checklist

- [ ] Formatted with `cargo fmt`
- [ ] Built with `nix build`
- [ ] Ran flake checks with `nix flake check`
- [ ] Added or updated relevant tests (leave unchecked if not applicable)
- [ ] Added or updated relevant documentation (leave unchecked if not applicable)
- [ ] Linked to related issues (leave unchecked if not applicable)

##### Validating with `install.determinate.systems`

If a maintainer has added the `upload to s3` label to this PR, it will become available for installation via `install.determinate.systems`:

```shell
curl --proto '=https' --tlsv1.2 -sSf -L https://install.determinate.systems/nix/pr/$PR_NUMBER | sh -s -- install
```

### CODE REVIEW

File: nix/tests/vm-test/default.nix [hunk 1]
```diff
@@ -4,7 +4,7 @@
 let
   nix-installer-install = ''
     NIX_PATH=$(readlink -f nix.tar.xz)
-    RUST_BACKTRACE="full" ./nix-installer install --nix-package-url "file://$NIX_PATH" --no-confirm --logger pretty --log-directive nix_installer=trace
+    RUST_BACKTRACE="full" ./nix-installer install --nix-package-url "file://$NIX_PATH" --no-confirm --logger pretty --log-directive nix_installer=info
```
Comment (Hoverbear) [inline nix/tests/vm-test/default.nix:7]: Ooops

File: src/action/linux/provision_selinux.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+        execute_command(Command::new("semodule").arg("-i").arg(&self.policy_path))
+            .await
+            .map_err(Self::error)?;
+
+        execute_command(Command::new("restorecon").args(["-FR", "/nix"]))
+            .await
+            .map_err(Self::error)?;
+
+        // execute_command(Command::new("systemctl").arg("daemon-reexec"))
```
Comment (Hoverbear) [inline src/action/linux/provision_selinux.rs:89]: Remove

File: src/planner/linux.rs [hunk 3]
```diff
@@ -139,26 +164,20 @@ fn check_not_wsl1() -> Result<(), PlannerError> {
...
-        // The device doesn't have SELinux set up
-        Err(e) if e.kind() == std::io::ErrorKind::NotFound => (),
-        // Some unknown error
-        Err(e) => {
-            tracing::warn!(error = ?e, "Got an error checking for SELinux setting, this install may fail if SELinux is set to `Enforcing`")
-        },
+async fn detect_selinux() -> Result<bool, PlannerError> {
+    if Path::new("/sys/fs/selinux").exists() {
+        // We expect systems with SELinux to have the normal SELinux tools.
+        // The policies for Nix are built as part of the installation process at this time.
```
Comment (Hoverbear) [inline src/planner/linux.rs:170]: Inaccurate

File: src/planner/mod.rs [hunk 4]
```diff
@@ -367,13 +367,8 @@ pub enum PlannerError {
...
     RosettaDetected,
     /// A Linux SELinux related error
-    #[error("\
-        This installer doesn't yet support SELinux in `Enforcing` mode.\n
-        \n\
-        If desirable, consider setting SELinux to `Permissive` mode with `setenforce Permissive`.\n\
-        \n\
-        If SELinux is important to you, please see https://github.com/DeterminateSystems/nix-installer/issues/124.")]
-    SelinuxEnforcing,
+    #[error("Unable to install on an SELinux system without common SELinux tooling, the binaries `restorecon`, `semodule_package`, and `semodule` are required")]
```
Comment (Hoverbear) [inline src/planner/mod.rs:370]: Inaccurate

File: src/action/common/configure_init_service.rs [hunk 5]
```diff
@@ -311,20 +317,6 @@ impl Action for ConfigureInitService {
                     })
                     .map_err(Self::error)?;
 
-                Self::check_if_systemd_unit_exists(SOCKET_SRC, SOCKET_DEST)
```
Comment (cole-h) [inline src/action/common/configure_init_service.rs:314]: Oh? Why was this removed?

File: src/action/linux/provision_selinux.rs [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                .map_err(|e| ActionErrorKind::CreateDirectory(parent.into(), e))
+                .map_err(Self::error)?;
+        }
+
+        tokio::fs::write(&self.policy_path, SE_LINUX_POLICY_PP_CONTENT)
+            .await
+            .map_err(|e| ActionErrorKind::Write(self.policy_path.clone(), e))
+            .map_err(Self::error)?;
+
+        execute_command(Command::new("semodule").arg("-i").arg(&self.policy_path))
```
Comment (cole-h) [inline src/action/linux/provision_selinux.rs:81]: Nit:

Do these SELinux binaries provide long flags for these options? It would be nice to see them instead, if possible, rather than the opaque `-i`, `-r`, etc.

File: src/action/linux/selinux/build.sh [hunk 7]
```diff
@@ -0,0 +1,5 @@
+#! /usr/bin/env nix-shell
+#! nix-shell -i bash -p libselinux -p semodule-utils -p checkpolicy
```
Comment (cole-h) [inline src/action/linux/selinux/build.sh:2]: We should probably pin this using our flake rather than taking the packages from whatever the user's nixpkgs is set to.

And what are the chances we could build this as part of the build process rather than checking binary files into the repo?

File: src/action/linux/selinux/nix.mod
Comment (cole-h) [inline src/action/linux/selinux/nix.mod:1]: Do we need this file if we only include the `.pp` bytes?

File: src/action/linux/selinux/build.sh (hunk 7 above)
Comment (Hoverbear) [reply src/action/linux/selinux/build.sh:2]: `nix-installer` is also a valid Rust tool, `cargo install nix-installer`. No way to do it. :(

File: src/action/linux/selinux/nix.mod
Comment (Hoverbear) [reply src/action/linux/selinux/nix.mod:1]: I think not!

File: src/action/linux/provision_selinux.rs (hunk 6 above)
Comment (Hoverbear) [reply src/action/linux/provision_selinux.rs:81]: You're right!

File: src/action/linux/selinux/build.sh (hunk 7 above)
Comment (cole-h) [reply src/action/linux/selinux/build.sh:2]: Right. RIP.

File: src/action/common/configure_init_service.rs (hunk 5 above)
Comment (Hoverbear) [reply src/action/common/configure_init_service.rs:314]: Correct me if I'm going insane, but it's a repeat of the above block and likely a bad merge?

File: src/action/linux/provision_selinux.rs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+use crate::action::{ActionError, ActionErrorKind, ActionTag};
+use crate::execute_command;
+
+use crate::action::{Action, ActionDescription, StatefulAction};
+
+const SE_LINUX_POLICY_PP_CONTENT: &[u8] = include_bytes!("selinux/nix.pp");
+
+/**
+Delete an operating system level user
```
Comment (Hoverbear) [inline src/action/linux/provision_selinux.rs:15]: I see a stale comment

File: src/action/common/configure_init_service.rs (hunk 5 above)
Comment (cole-h) [reply src/action/common/configure_init_service.rs:314]: This block was for `SOCKET_{SRC,DEST}`, while the block above is for `SERVICE_{SRC,DEST}`, so I don't think it's a repeat? But I could be wrong.

Comment (Hoverbear) [reply src/action/common/configure_init_service.rs:314]: Oh my god.

**Comment by Hoverbear** [comment]
https://github.com/DeterminateSystems/nix-installer/assets/130903/2cb82959-eea9-4171-8807-a71a48abdc2e

File: src/action/linux/selinux/nix.pp
Comment (cole-h) [inline src/action/linux/selinux/nix.pp:1]: Is this portable to various SELinux systems?

Basically wondering if it sidesteps the issues that e.g. dynamic linking has.

Comment (Hoverbear) [reply src/action/linux/selinux/nix.pp:1]: I am not sure. I tested this over Fedora 36 and 7, and RHEL 7 8 & 9. It worked ok.

**Review by cole-h — APPROVED** [review]

File: src/action/linux/selinux/nix.pp
Comment (cole-h) [reply src/action/linux/selinux/nix.pp:1]: I suppose if people have issues, we'll be the first to know :P

**Review by lheckemann — APPROVED** [review]
